introduce definitely_contiguous  and use it for reshape and tensor meta data computation.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #154388
* #153441
* __->__ #153432

when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
known _contig

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ColinPeppler commented on torch/_prims_common/__init__.py:
```diff
@@ -362,6 +432,25 @@ def is_contiguous_for_memory_format(  # type: ignore[return]
+    if memory_format == torch.channels_last_3d:
+        return is_known_channels_last_contiguous_3d(a)
+
+    torch._check(
+        False,
+        lambda: f"is_contiguous received unsupported memory format {memory_format}",
+    )
```
would it be better to raise an exception instead of torch._check(false)?

laithsakka replied:
mm i just copied 
```
def is_contiguous_for_memory_format(  # type: ignore[return]
    a: Tensor, *, memory_format: torch.memory_format
) -> bool:
    validate_memory_format(memory_format)
    if memory_format == torch.contiguous_format:
        return is_contiguous(a)
    if memory_format == torch.channels_last:
        return is_channels_last_contiguous_2d(a)
    if memory_format == torch.channels_last_3d:
        return is_channels_last_contiguous_3d(a)
    torch._check(
        False,
        lambda: f"is_contiguous received unsupported memory format {memory_format}",
    )
    ```
    I will do a refatctor as mentioned above to pass a flag instead of replicating the function

laithsakka replied:
this is existing code, with the new change it will show more appropriately

ColinPeppler commented on torch/_prims_common/__init__.py:
```diff
@@ -329,6 +376,29 @@ def is_channels_last_contiguous_3d(a: Tensor) -> bool:
 
 
+# similar to is_channels_last_contiguous_3d but return false on data dependency.
+def is_known_channels_last_contiguous_3d(a: Tensor) -> bool:
```
this one feel like it could re-use code w/ is_known_channels_last_contiguous_2d

laithsakka replied:
what i did is that i copied exisiting functions and used guard_or_false in those guard_or_true in those new functions 

example: this is a copy of same for all others
```
def is_channels_last_contiguous_3d(a: Tensor) -> bool:
    # NDHWC or not channels last 3D contiguous
    if a.ndim != 5:
        return False
    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
    expected_stride = 1
    for idx in (1, 4, 3, 2, 0):
        length = a.shape[idx]
        if guard_size_oblivious(length == 1):
            continue
        stride = a.stride()[idx]
        if guard_size_oblivious(stride != expected_stride):
            return False
        expected_stride *= length
    return True
    ```

laithsakka replied:
mm maybe not actually it looked ugly when i did that. let me think

ColinPeppler commented on torch/_prims_common/__init__.py:
```diff
@@ -258,6 +258,30 @@ def check_all_strides(
+    """
+    from torch.fx.experimental.symbolic_shapes import guard_or_false, guard_or_true
+
+    if guard_or_false(a.numel() == 0):
```
can this be more generalized?

https://github.com/pytorch/pytorch/blob/dde705864a25991fead47a10e965bccb4a054b9c/torch/_prims_common/__init__.py#L271-L272

laithsakka replied:
it can it wont make a difference logic is the same.

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce is_known _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use is_known _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if is_known _contiguous  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented on c10/core/Contiguity.h:
```diff
@@ -12,24 +12,49 @@ namespace c10 {
 template <typename T>
 bool _compute_contiguous(ArrayRef<T> sizes, ArrayRef<T> strides, T numel) {
-  bool is_contiguous = true;
   if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(numel, 0))) {
```
unifying python and c++ impls with respect to guarding and simplifying the loop

laithsakka commented on test/test_dynamic_shapes.py:
```diff
@@ -3249,6 +3249,31 @@ def forward(self, arg0_1: "i64[2][1]cpu", arg1_1: "Sym(u2)", arg2_1: "Sym(u3)",
+        with ctx():
+            # This used to hit could guard on data-dependent expression Eq(10, u3) x.stride[0]==10. and x.size()=[u2, u3].
+            # but not anymore since we use is_known_contiguous.
+            # We need a way to mark trides unbacked to avoid the recompilation here.
```
trides -> strides.

laithsakka commented on torch/fx/passes/shape_prop.py:
```diff
@@ -32,6 +33,10 @@ class TensorMetadata(NamedTuple):
 
 
+# When include_contiguity is True, we will set contiguity when its always true for the tensor.
+# some tensors can rerepsent both contiguous and non contiguous tensors. Ex (u0, u1) with (u2, u3).
```
. Ex (u0, u1) with (u2, u3).-> ex a tensor of size (u0, u1) and (u2, u3) strides.

bobrenjc93 replied:
s/rerepsent/represent

bobrenjc93 commented on c10/core/Contiguity.h:
```diff
@@ -12,24 +12,49 @@ namespace c10 {
+}
+
+// This function will return True if the tensor is contiguous, and False if the
+// its not or if we can't determine if it is contiguous due to unbacked symbols
```
s/its/it's

bobrenjc93 commented on test/test_dynamic_shapes.py:
```diff
@@ -3249,6 +3249,31 @@ def forward(self, arg0_1: "i64[2][1]cpu", arg1_1: "Sym(u2)", arg2_1: "Sym(u3)",
+        log_stream, ctx = logs_to_string(
+            "torch._functorch._aot_autograd.dispatch_and_compile_graph", "aot_graphs"
+        )
+        with ctx():
```
is this logging for debugging or you meant to check it in?

laithsakka replied:
i meant to check it in .

bobrenjc93 replied:
but you don't seem to use log_stream anywhere?

bobrenjc93 commented on torch/_prims_common/__init__.py:
```diff
@@ -259,69 +259,90 @@ def check_all_strides(
 
 # This function is equivalent to compute_contiguous() from TensorImpl.cpp
-def is_contiguous(a: TensorLikeType) -> bool:
+def is_contiguous(a: TensorLikeType, false_if_unknown=False) -> bool:
```
hmm I'm not a big fan of this fork. Do you have known cases where we must have false_if_unknown=False? If so, can you please add a comment justifying the need for this complexity?

laithsakka replied:
the idea is to avoid copying each one of those functions twice and replicate the logic, 
we cant assume all existing calls to is_contiguous have the semantics of is_known_contiguous.
we have to check them all one by one and move them as needed only if they are not material. 

ditto for all others.

laithsakka replied:
this is one usage that I am addressing in the next PR
https://github.com/pytorch/pytorch/pull/153432
let me see what others we have, 
but in general, if you want to know if something is contiguous or not and care about the result 100% you will call 
is_contiguous_for_memory_format else if you only want best effort and its always safe if the function return false you will call is_known_contiguous_for_memory_format

laithsakka replied:
here is other usages:
(1) i see them used in class SymNodes. I am not sure if they are called from c++. 
pytorch/torch/fx/experimental/sym_node.py

(2) The function 
def is_non_overlapping_and_dense(a: Tensor) -> bool:
calls 
```
 if is_contiguous(a) or is_channels_last_contiguous(a):
        return True
 ```

ColinPeppler replied:
> we cant assume all existing calls to is_contiguous have the semantics of is_known_contiguous.

IIUC, it sounds like sometimes we want a data-dependent exception to force the user to add a hint, so that the make an important call on how the program is compiled. On the other hand, in more lax cases, we can just use a best guess i.e. guard_or_f/t.

What if we renamed `false_if_unknown` to emphasize this like `relaxed=False`?

bobrenjc93 replied:
Instead of forking, can we just introduce a new api `definitely_contiguous` and use that instead? This fork is really confusing.

laithsakka replied:
i had that in the first is a lot of code duplication that need to be kept in sync. i like this more see this comment.

bobrenjc93 commented on torch/_prims_common/__init__.py:
```diff
@@ -259,69 +259,90 @@ def check_all_strides(
 
 # This function is equivalent to compute_channels_last_contiguous_2d() in TensorImpl.cpp
-def is_channels_last_contiguous_2d(a: Tensor) -> bool:
+def is_channels_last_contiguous_2d(a: Tensor, false_if_unknown=False) -> bool:
```
ditto

bobrenjc93 commented on torch/_prims_common/__init__.py:
```diff
@@ -259,69 +259,90 @@ def check_all_strides(
 
 
-def is_channels_last_contiguous_3d(a: Tensor) -> bool:
+def is_channels_last_contiguous_3d(a: Tensor, false_if_unknown=False) -> bool:
```
ditto

bobrenjc93 commented on torch/_prims_common/__init__.py:
```diff
@@ -345,23 +366,46 @@ def validate_memory_format(memory_format: torch.memory_format):
 
 def is_contiguous_for_memory_format(  # type: ignore[return]
-    a: Tensor, *, memory_format: torch.memory_format
+    a: Tensor, *, memory_format: torch.memory_format, false_if_unknown=False
```
ditto

eellison commented:
We should error we then later mutate this tensor or its alias, because then we are potentially diverging from eager. including if this gets output in user programs.

laithsakka commented:
> We should error we then later mutate this tensor or its alias, because then we are potentially diverging from eager. including if this gets output in user programs.

is this comment meant for the reshape PR?

bobrenjc93 reviewed (approved):
We can always debate/rename/refactor, overall I think this is a step in the right direction. Please clean up the nits before landing though.

laithsakka committed:
Update on "introduce definitely _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce definitely _contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce definitely_contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce definitely_contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "introduce definitely_contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "introduce definitely_contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

pianpwk commented on torch/_refs/__init__.py:
```diff
@@ -3824,7 +3825,7 @@ def _view_simple(a: TensorLikeType, shape, data_dependent_error) -> TensorLikeTy
 
-    if a.is_contiguous():
+    if definitely_contiguous(a):
         return a.as_strided(shape, utils.make_contiguous_strides_for(shape))
```
also apply this to https://github.com/pytorch/pytorch/blob/b8452e55bc50109a14add60397fee6c85486ae8d/torch/_refs/__init__.py#L3981 ?

laithsakka committed:
Update on "introduce definitely_contiguous  and use it for reshape and tensor meta data computation. "


when a tensor has unbacked symbols it can be general enough to represent both contiguous and non contiguous tensors. 
in that case we cant really evaluate is_contiguous. In many places in the code base, we check for is_contiguous to take a fast path. but the general path usually works for both contiguous and not contiguous in that case we probably want 
to use definitely _contiguous API. 

This is appleid for reshape in this PR and also to  tensor meta data computation, the meta data now will have an attribute that says that its contiguous when its always contiguous. We would store that only if definitely _contiguous is true  now. 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
